test_project/adam.py

```diff
@@ -101,3 +101,13 @@ def single_iteration():
         result = True
 
     return result
+
+
+def raise_exception():
```

Comment:
I think it would be clearer if this code just raised an exception. Then the test could look for that exception. It would make the code more concise and let it "speak for itself" without the reader needing to understand why values are returned, etc.